### PR TITLE
MAINT: Fix more typing and enable mypy on more files.

### PR DESCRIPTION
**Pull request by Carreau**
Step toward #2751

Start taking mypy errors one by one and fix/ignore them.

### CODE REVIEW

File: napari/_app_model/constants/_commands.py [hunk 1]
```diff
@@ -77,7 +77,7 @@ class CommandId(str, Enum):
     LAYER_PROJECT_MEDIAN = 'napari:layer:project_median'
 
     @property
-    def title(self) -> str:
+    def title(self) -> str: # type: ignore[override]
```
Comment (Carreau) [inline napari/_app_model/constants/_commands.py]: str, already have a title method, so we're shadowing it with a property which is technically incompatible. So ignore.

File: napari/_qt/_qapp_model/_menus.py [hunk 2]
```diff
@@ -3,7 +3,7 @@
 from app_model.backends.qt import QModelMenu
 
 if TYPE_CHECKING:
-    from qtpy.QtWidgets import QWidget
+    from qtpy.QtWidgets import QWidget  # type: ignore[attr-defined]
```
Comment (Carreau) [inline napari/_qt/_qapp_model/_menus.py:6]: All those exists, but are can't be seen by mypy (yet?)

File: napari/_qt/menus/window_menu.py [hunk 3]
```diff
@@ -1,14 +1,15 @@
...
 if TYPE_CHECKING:
+    from napari._qt.menus._util import MenuItem
     from napari._qt.qt_main_window import Window
 
 
 class WindowMenu(NapariMenu):
     def __init__(self, window: 'Window') -> None:
         super().__init__(trans._('&Window'), window._qt_window)
-        ACTIONS = []
+        ACTIONS: List[MenuItem] = []
```
Comment (Carreau) [inline napari/_qt/menus/window_menu.py:14]:
```suggestion
        ACTIONS: List['MenuItem'] = []
```

File: napari/plugins/npe2api.py [hunk 4]
```diff
@@ -88,21 +91,25 @@ def iter_napari_plugin_info() -> Iterator[Tuple[PackageMetadata, bool, dict]]:
...
+        info_copy.pop("display_name", None)
+        pypi_versions = info_copy.pop("pypi_versions")
+        conda_versions = info_copy.pop("conda_versions")
+        info_ = cast(_ShortSummaryDict, info_copy)
 
         # TODO: use this better.
         # this would require changing the api that qt_plugin_dialog expects to
         # receive
-        info_.pop("display_name", None)
+        # assert 'display_name' not in info_
```
Comment (Carreau) [inline napari/plugins/npe2api.py]:
```suggestion
```

File: napari/plugins/npe2api.py [hunk 5]
```diff
@@ -88,21 +91,25 @@ def iter_napari_plugin_info() -> Iterator[Tuple[PackageMetadata, bool, dict]]:
...
         extra_info = {
             'home_page': info_.get("home_page", ""),
-            'pypi_versions': info_.pop("pypi_versions"),
-            'conda_versions': info_.pop("conda_versions"),
+            'pypi_versions': pypi_versions,
+            'conda_versions': conda_versions,
         }
         info_["name"] = normalized_name(info_["name"])
-        meta = PackageMetadata(**info_)
+        meta = PackageMetadata(**info_)  # type: ignore [call-arg]
```
Comment (Carreau) [inline napari/plugins/npe2api.py]: This complains about all the missing attributes.

File: pyproject.toml [hunk 6]
```diff
@@ -394,7 +392,10 @@ module = [
     'napari.layers.utils.string_encoding',
     'napari.layers.utils.style_encoding',
     'napari.layers.utils.text_manager',
+    'napari.plugins._npe2',
+    'napari.plugins._npe2',
```
Comment (Carreau) [inline pyproject.toml]:
```suggestion
```

**Review by Czaki — APPROVED** [review]

File: napari/_qt/menus/window_menu.py [hunk 7]
```diff
@@ -1,14 +1,15 @@
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, List
```
Comment (melonora) [inline napari/_qt/menus/window_menu.py:1]: @Czaki Do you know when we would move to drop 3.8 support? Would that be with 0.5.0, if so perhaps better to switch to `list` already here.

File: napari/_app_model/constants/_commands.py [hunk 8]
```diff
@@ -76,7 +76,7 @@ class CommandId(str, Enum):
...
```
Comment (melonora) [inline napari/_app_model/constants/_commands.py]:
```suggestion
    def title(self) -> str:  # type: ignore[override]
```
Has 1 space, switch to conventional 2 spaces.
I see mypy indeed does not recognize that it means to override, which would lead an error. I had a look to see what would change if we didn't have this as property, it leads to...other type errors in `ToggleViewerAction`. Could be that we need to have a look, but not part of this PR.

Comment (melonora) [inline napari/_app_model/constants/_commands.py]:
```suggestion
    def title(self) -> str:  # type: ignore[override]
```
Has 1 space, switch to conventional 2 spaces.
I see mypy indeed does not recognize that it means to override, which would lead an error. I had a look to see what would change if we didn't have this as property, it leads to...other type errors in ToggleViewerAction. Could be that we need to have a look, but not part of this PR.

File: napari/_app_model/_app.py [hunk 9]
```diff
@@ -38,8 +38,8 @@ def __init__(self, app_name=APP_NAME) -> None:
         self.menus.append_menu_items(SUBMENUS)
 
     @classmethod
-    def get_app(cls) -> NapariApplication:
-        return Application.get_app(APP_NAME) or cls()
+    def get_app(cls) -> NapariApplication:  # type: ignore[override]
+        return Application.get_app(APP_NAME) or cls()  # type: ignore[return-value]
```
Comment (melonora) [inline napari/_app_model/_app.py]:
```suggestion
    def get_app(cls, app_name: str = APP_NAME) -> NapariApplication:  
        return Application.get_app(app_name) or cls()  # type: ignore[return-value]
```
I think this would be a small fix matching the signature of the `class_method` `get_app` in `Application` and thereby omitting the need for the ignore

File: napari/_qt/_qapp_model/_menus.py (hunk 2 above)
Comment (melonora) [reply napari/_qt/_qapp_model/_menus.py:6]: Hmm I tried without and it actually did not give me problems when running:
`mypy  --python-version=3.x .\napari\_qt\_qapp_model\_menus.py` with x being 8, 9 or 10

File: napari/_qt/containers/_base_item_model.py [hunk 10]
```diff
@@ -12,7 +12,7 @@
 if TYPE_CHECKING:
     from typing import Optional
 
-    from qtpy.QtWidgets import QWidget
+    from qtpy.QtWidgets import QWidget  # type: ignore[attr-defined]
```
Comment (melonora) [inline napari/_qt/containers/_base_item_model.py:15]: same here

File: napari/_qt/containers/qt_list_view.py [hunk 11]
```diff
@@ -10,7 +10,7 @@
 if TYPE_CHECKING:
     from typing import Optional
 
-    from qtpy.QtWidgets import QWidget
+    from qtpy.QtWidgets import QWidget  # type: ignore[attr-defined]
```
Comment (melonora) [inline napari/_qt/containers/qt_list_view.py:13]: and here

File: napari/_app_model/_app.py (hunk 9 above)
Comment (melonora) [inline napari/_app_model/_app.py]:
```suggestion
    def get_app(cls, app_name: str = APP_NAME) -> NapariApplication: 
        return Application.get_app(app_name) or cls()  # type: ignore[return-value]
```
I think this would be a small fix matching the signature of the class_method get_app in Application and thereby omitting the need for the ignore

File: napari/_qt/menus/window_menu.py (hunk 3 above)
Comment (melonora) [inline napari/_qt/menus/window_menu.py:14]: @Czaki is the plan to drop support for 3.8 in version 0.5.0? If so maybe change to `list` here.

**Review by melonora — CHANGES_REQUESTED** [review]
Thanks for the work! Mostly I have minor questions, but I would like the fix implemented regarding `APP_NAME`, unless @Czaki disagrees.

File: napari/_qt/menus/window_menu.py (hunk 7 above)
Comment (Czaki) [reply napari/_qt/menus/window_menu.py:1]: I'm not sure what is conclussion here: https://napari.zulipchat.com/#narrow/stream/212875-general/topic/NEP29.20misunderstandings.20and.20misapplications 

But I think that change minimum python version in ruff configuration will replace `List` with `list`

File: napari/_app_model/constants/_commands.py (hunk 8 above)
Comment (Czaki) [reply napari/_app_model/constants/_commands.py]: Strange that black do not catch this. Based on this https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#comments it should be fixed by black in pre-commit.

File: napari/_qt/_qapp_model/_menus.py (hunk 2 above)
Comment (Czaki) [reply napari/_qt/_qapp_model/_menus.py:6]: I observe in past that mypy on my machine returns different result that one on github actions, even when running `tox -e mypy` 
When you run mypy from dev env it may use all of your libraries. 

@Carreau it does not exist for mypy as we do not install any Qt backend in mypy env:
https://github.com/napari/napari/blob/main/resources/requirements_mypy.in

Maybe we should add pyqt5 (or 6) to list of requirements

File: napari/_qt/menus/window_menu.py (hunk 7 above)
Comment (Carreau) [reply napari/_qt/menus/window_menu.py:1]: See https://github.com/napari/napari/pull/5768 for dropping 3.8

File: napari/_qt/_qapp_model/_menus.py (hunk 2 above)
Comment (Carreau) [reply napari/_qt/_qapp_model/_menus.py:6]: Ok, I can get something out of adding pyqt6 as a dependency. Can we postpone that to another PR though ? It triggers a bunch of other type errors that I'd prefer to deal with later.

File: napari/_qt/menus/window_menu.py (hunk 7 above)
Comment (Czaki) [reply napari/_qt/menus/window_menu.py:1]: I put topic of python 3.8 drop on today community meeting.

File: napari/_qt/_qapp_model/_menus.py (hunk 2 above)
Comment (Czaki) [reply napari/_qt/_qapp_model/_menus.py:6]: Yes. I think that it will interact with so many modules that it need separate PR.

Comment (melonora) [reply napari/_qt/_qapp_model/_menus.py:6]: I agree as well.

File: napari/_qt/menus/window_menu.py (hunk 7 above)
Comment (jni) [reply napari/_qt/menus/window_menu.py:1]: The overall feeling in the meeting was that the improved legibility of typing was worth the drop despite the discussion on Zulip.

Comment (Czaki) [reply napari/_qt/menus/window_menu.py:1]: Yes. We will wait on 0.4.19 release (to simplify backport) and then drop 3.8.

**Review by melonora — APPROVED** [review]
Seems good to go to me!

**Comment by Carreau** [comment]
thanks!
